crates/codegen/parser/runtime/src/cst.rs

```diff
@@ -46,63 +46,103 @@ impl Node {
         }
     }
 
-    pub fn create_cursor(&self, text_offset: TextIndex) -> Cursor {
+    /// Creates a [`Cursor`] that starts at the current node as the root and a given initial `text_offset`.
+    pub fn cursor_with_offset(&self, text_offset: TextIndex) -> Cursor {
         Cursor::new(self.clone(), text_offset)
     }
 
+    /// Reconstructs the original source code from the parse tree.
+    pub fn reconstruct(self) -> String {
+        match self {
+            Self::Rule(rule) => rule.reconstruct(),
+            Self::Token(token) => token.text.clone(),
+        }
+    }
+
     pub fn as_rule(&self) -> Option<&Rc<RuleNode>> {
         match self {
             Self::Rule(node) => Some(node),
             _ => None,
         }
     }
 
+    pub fn into_rule(self) -> Option<Rc<RuleNode>> {
```

Comment:
I wonder why `into_rule()` in addition to `as_rule()`? What new use case does it enable?

I don't see any current usage/invocation. Is this intended?

same question for `into_token()`.

Reply:
`into` should always be for a construction, and `as` for a view.

Reply:
In that case, shouldn't we use `impl From` for these conversions?
This will enable `let rule = node.into()`, and also `Rule::from(node)` automatically.

Reply:
It's not `Rule` that we're creating, it's specifically `Option<Rc<Rule>>`.

I'm not sure if we can implement `From<Node> for Option<Rc<Rule>>` due to orphan rules, but maybe (since the `Node` type is defined by us)?

However, again, like with `TryFrom`, is there concrete value from these impls?

We'd have to use it as `Option::<Rc<Rule>>::from(node)` at call-site or `let rule: Option<Rc<Rule>> = node.into();` (ditto for a separate `From<&Node> for Option<&Rc<Rule>>` impl).

These are valuable when used in a generic setting or when the type is given at some position (so the trait implementer can be inferred, especially for a mouthful like `Option<Rc<Rule>>`) and I'm not sure there are really use-cases for these at the moment.